### PR TITLE
Access Control: Add rolepicker when Fine Grained Access Control is enabled

### PR DESCRIPTION
**Pull request by briangann**
**What this PR does / why we need it**:

This PR adds role picker to the "Add user to organization" modal if access control enabled (instead of default picker).

Several changes were made to existing components to allow them to behave more intuitively:

* When the "Add User To Organization" button is used, the modal will not allow selecting roles before selecting an Org
* Orgs that a user is already a member of are not displayed in the modal
* Roles in the modal will show an "Apply" button, which will be passed back to the parent
* The table mode for editing an existing user, the role menu shows  "Update" and commits the changes

This PR also adds role configurations for "Service Accounts"

**Which issue(s) this PR fixes**:

This builds off/replaces PR https://github.com/grafana/grafana/pull/44271

### CODE REVIEW

File: public/app/features/serviceaccounts/ServiceAccountsListPage.tsx
Comment (alexanderzobnin) [inline public/app/features/serviceaccounts/ServiceAccountsListPage.tsx:77]: Shouldn't it be with `await`? Otherwise there might be a race condition and fetchServiceAccounts may return old result.

File: public/app/features/serviceaccounts/ServiceAccountCreatePage.tsx
Comment (eleijonmarck) [inline public/app/features/serviceaccounts/ServiceAccountCreatePage.tsx:92]: aha, so you create a service account and within the request you use the update api to update the roles. 🤔 

Interesting. I was working on making the role apart of the creation api. But that does not need to happen. This just creates two queries instead.

nice approach

File: public/app/core/components/RolePicker/RolePickerMenu.tsx
Comment (eleijonmarck) [inline public/app/core/components/RolePicker/RolePickerMenu.tsx:176]: I guess we need the UIDs for this to work?

if not; we could remove selectedCustomRoles entirely

File: public/app/features/serviceaccounts/ServiceAccountsListPage.tsx
Comment (briangann) [reply public/app/features/serviceaccounts/ServiceAccountsListPage.tsx:77]: good catch! will fix that

File: public/app/features/admin/UserOrgs.tsx
Comment (eleijonmarck) [inline public/app/features/admin/UserOrgs.tsx:338]: should be removed

File: public/app/features/serviceaccounts/ServiceAccountCreatePage.tsx
Comment (eleijonmarck) [inline public/app/features/serviceaccounts/ServiceAccountCreatePage.tsx:86]: nit: maybe this is missed or here on purpose, not sure. but just wanted to make sure that we clean potential logging.

**Comment by eleijonmarck** [comment]
@briangann let me know if there is any questions to my comments. One thing is the approach of service accounts:
1. creating the service account
2. updating the service account with the role

If any one of the request fail, we cannot roll back. So it will be a created service account without a role attached

File: public/app/features/serviceaccounts/ServiceAccountCreatePage.tsx
Comment (briangann) [reply public/app/features/serviceaccounts/ServiceAccountCreatePage.tsx:86]: was thinking of adding error responses vs blinding failing, left it in as a TODO

**Comment by briangann** [comment]
> @briangann let me know if there is any questions to my comments. One thing is the approach of service accounts:
> 
> 1. creating the service account
> 2. updating the service account with the role
> 
> If the first request fail, we cannot roll back. So it will be a created service account without a role attached

Understood, i'll add in logic so it doesn't continue to try to update builtin roles and setup FGAC roles.

On create failure it should just error out (and also show errors if the other two steps fail)

File: public/app/core/components/Select/OrgPicker.tsx
Comment (Clarity-89) [inline public/app/core/components/Select/OrgPicker.tsx:34]:
```suggestion
        return !idArray.includes(item.value.id);
```

**Review by Clarity-89 — COMMENTED** [review]
Haven't tested it, but the code looks good in general. Left a few minor comments.

**Comment by vtorosyan** [comment]
I am going to test this today / tomorrow and can confirm the functional part afterwards.

File: public/app/features/admin/UserOrgs.tsx
Comment (eleijonmarck) [inline public/app/features/admin/UserOrgs.tsx:166]: what happens if this first request fails?

wondering if it is better to implement to remove the permissions within the same api call to avoid bad states, where one of these api calls fails.
@briangann wdyt?

@kalleep maybe you have something better in mind 🤔

**Comment by vtorosyan** [comment]
@briangann FYI I had to merge main branch to test with the up to date fixes. Sorry for no in advance notice.

File: public/app/features/admin/UserOrgs.tsx
Comment (briangann) [reply public/app/features/admin/UserOrgs.tsx:166]: general error handling needs work, previously it would leave roles behind for deleted accounts too, and when you went to add them back in, the old roles appeared "out of nowhere"

**Review by vtorosyan — APPROVED** [review]
I have done functional manual testing and the change looks good and solved the problem. Thanks @briangann for the work. Note that I am accepting just to indicate that functionally changes look good to go - would be great if someone with more expertise in the code can also +1 before we merge.

File: public/app/features/serviceaccounts/ServiceAccountCreatePage.tsx
Comment (briangann) [reply public/app/features/serviceaccounts/ServiceAccountCreatePage.tsx:80]: pretty certain there were additional fields so I just pulled out what was needed here.

**Comment by vtorosyan** [comment]
We have discussed the changes with @briangann and agreed to merge this PR it is generally ready for the Grafana 9.0 beta release. The error handling improvements are still planned in a separate PR, which is not a blocker for this to be shipped. If there are any concerns, please follow up on this PR so that we can handle it post merge.

**Comment by gfl3** [comment]
unsubscribe

**Comment by rthomasson** [comment]
Unsubscribe

On Jun 3, 2022, at 4:30 PM, gfl3 ***@***.***> wrote:

﻿

unsubscribe

—
Reply to this email directly, view it on GitHub<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fgrafana%2Fgrafana%2Fpull%2F48347%23issuecomment-1146337066&data=05%7C01%7Crthomasson%40cassidy.com%7Cf2fbc5ee14f4462ee39608da459fd4a2%7C54247946c4cc4f10a9449656acacb39b%7C0%7C0%7C637898850026614175%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=L8W4ynsn8aBQ%2FYWfDPdIYUEJYkbT8odheQ8CrfGG8PY%3D&reserved=0>, or unsubscribe<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAAJA6TLRJQKUPW24ZJ57FBTVNJTLFANCNFSM5UPBF5JA&data=05%7C01%7Crthomasson%40cassidy.com%7Cf2fbc5ee14f4462ee39608da459fd4a2%7C54247946c4cc4f10a9449656acacb39b%7C0%7C0%7C637898850026614175%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=6cyblTIZ4uPrQfRGrIrUMqwxC6ztApyYvuFrHLT6ug4%3D&reserved=0>.
You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

CAUTION: This email originated from outside the organization. Do not click links or open attachments unless you recognize the sender and know the content is safe.

**Pull request merged**
